### PR TITLE
add experimental tracing using a native event collector

### PR DESCRIPTION
**Pull request by rochdev**
### What does this PR do?

Add experimental tracing using a native event collector that stores and processes traces outside of the application.

### Motivation

This was previously just a proof-of-concept. The goal of this PR is to have at last something on the main branch with proper CI tests so that we stay compatible with the use of events.

### CODE REVIEW

File: packages/dd-trace/src/collector/exporter.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+      [segmentDiscardChannel, encoder.encodeSegmentDiscard.bind(encoder)],
+      [spanStartChannel, encoder.encodeSpanStart.bind(encoder)],
+      [spanFinishChannel, encoder.encodeSpanFinish.bind(encoder)],
+      [tagsChannel, encoder.encodeAddTags.bind(encoder)],
+      [errorChannel, encoder.encodeException.bind(encoder)]
+    ])
+  }
+
+  start () {
+    this._handlers.forEach((cb, ch) => ch.subscribe(cb))
```
Comment (simon-id) [inline packages/dd-trace/src/collector/exporter.js:28]: nitpick: cb and ch are difficult to tell apart when reading fast

**Review by simon-id — COMMENTED** [review]
didn't do a *super* in depth review (especially for encoder.js), few nitpicks and questions but LGTM

File: packages/dd-trace/src/collector/exporter.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+const spanFinishChannel = channel('datadog:tracing:span:finish')
+const tagsChannel = channel('datadog:tracing:span:tags')
+const errorChannel = channel('datadog:tracing:span:error')
+
+class CollectorExporter {
+  constructor (config) {
+    const encoder = this._encoder = new Encoder(config)
+
+    this._handlers = new Map([
+      [segmentStartChannel, encoder.encodeSegmentStart.bind(encoder)],
```
[the remark this answers is not in the document]
Comment (rochdev) [reply packages/dd-trace/src/collector/exporter.js:18]: It happens only once when the exporter is created so that shouldn't matter (unless you mean the resulting function is slower).

[the remark this answers is not in the document]
Comment (rochdev) [reply packages/dd-trace/src/collector/exporter.js:18]: I couldn't find any evidence of this.

[the remark this answers is not in the document]
Comment (simon-id) [reply packages/dd-trace/src/collector/exporter.js:18]: yeah it looks like for the execution, it's the same ballpark, it's for the creation that it's slower

File: packages/dd-trace/src/collector/exporter.js (hunk 1 above)
Comment (rochdev) [reply packages/dd-trace/src/collector/exporter.js:28]: Is it fine to leave it as-is? While I understand the sentiment, it's just a very simple one-liner, so I prioritize keeping the line short when possible.

File: packages/dd-trace/src/collector/exporter.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (rochdev) [reply packages/dd-trace/src/collector/exporter.js:18]: So can I keep `bind`? Also, do you have any article or benchmark to share for future reference?

[the remark this answers is not in the document]
Comment (simon-id) [reply packages/dd-trace/src/collector/exporter.js:18]: Yes looks like it's fine :+1: I tried googling a little but nothing super up to date or non-contradictory with other articles, so I just tried on my own machine.
Results are that calling `.bind()` is way slower than creating a closure, but to execute them it's basically the same. With or without a `this` value, or bound arguments.

**Comment by simon-id** [comment]
comments closed, LGTM :+1:

File: packages/dd-trace/src/plugins/util/web.js [hunk 3]
```diff
@@ -461,19 +462,6 @@ function addResponseTags (context) {
...
-  const { req, span } = context
-  const tags = span.context()._tags
-
-  if (tags['resource.name']) return
-
-  const resource = [req.method, tags[HTTP_ROUTE]]
-    .filter(val => val)
-    .join(' ')
-
-  span.setTag(RESOURCE_NAME, resource)
```
Comment (szegedi) [inline packages/dd-trace/src/plugins/util/web.js:474]: Why is `RESOURCE_NAME` tag no longer added?

We're kinda-sorta using it [here](https://github.com/DataDog/dd-trace-js/blob/master/packages/dd-trace/src/profiling/profilers/wall.js#L35) for establishing endpoint names to associate with profiles, although we can also compute the value from `HTTP_METHOD` and `HTTP_ROUTE` tags when `RESOURCE_NAME` isn't present.

Comment (rochdev) [reply packages/dd-trace/src/plugins/util/web.js:474]: It's still set but later in the pipeline. At this point the resource name cannot be trusted anyway because we allow changing it up until the span is finished, which is often used by customers.

File: packages/dd-trace/src/collector/span.js [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    // TODO: Emit a discard event from tracer. For now we just short-circuit.
+    if (trace.isDiscarded) return
+    if (trace.isRecording === false) {
+      trace.isDiscarded = true
+      return segmentDiscardChannel.publish({ segmentId })
+    }
+
+    const ticks = finishTime
+      ? now() + finishTime - trace.ticks - trace.startTime
+      : now() - trace.ticks
```
Comment (Qard) [inline packages/dd-trace/src/collector/span.js:36]: Can we do this on the collector side? Otherwise we need to keep the timing state in the library. I feel like it _should_ be safe to just send a Date.now() in the start event and monotonic timestamps in each of the start and end, then compute the delta in the collector.

File: packages/dd-trace/src/collector/span.js [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+const startSegmentChannel = channel('datadog:tracing:segment:start')
+const segmentDiscardChannel = channel('datadog:tracing:segment:discard')
+const startChannel = channel('datadog:tracing:span:start')
+const tagsChannel = channel('datadog:tracing:span:tags')
+const errorChannel = channel('datadog:tracing:span:error')
+const finishChannel = channel('datadog:tracing:span:finish')
+
+let segmentId = 0
+
+class DatadogCollectorSpan extends DatadogSpan {
```
Comment (Qard) [inline packages/dd-trace/src/collector/span.js:18]: This sub-classing is rather difficult to follow. Can we not just put all this logic in-place in `DatadogSpan` and use `channel.hasSubscribers` to skip it? There's a lot of circular dependency between this and the parent class calling methods on this that makes it challenging to understand the flow. I feel like it'd be a lot easier to understand if it was done in the parent class instead, or even with plain functions _called_ from the parent class. 🤔

File: packages/dd-trace/src/collector/span.js [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    if (keyValuePairs.error) {
+      errorChannel.publish({
+        segmentId: this._spanContext._trace.segmentId,
+        spanIndex: this._spanContext._spanIndex,
+        error: keyValuePairs.error
+      })
+    }
+  }
+
+  _trackSegment () {
```
Comment (Qard) [inline packages/dd-trace/src/collector/span.js:92]: Why do spans need awareness of segments now? Can we avoid this somehow and leave that up to the collector?

File: packages/dd-trace/src/collector/span.js (hunk 4 above)
Comment (rochdev) [reply packages/dd-trace/src/collector/span.js:36]: That's basically what this is doing except that it does that at the trace level instead of the span level. This is for efficiency but also because it was otherwise causing intermittent slight drifts. It's also worth noting that the computation is only happening when a finish time if provided, otherwise it only sends the monotonic timestamp.

File: packages/dd-trace/src/collector/span.js (hunk 5 above)
Comment (rochdev) [reply packages/dd-trace/src/collector/span.js:18]: The idea was to isolate the logic to a separate class instead of mixing it with the existing logic. I could instead just put all the code in the same place but it seemed like it would be even _more_ confusing. I felt like keeping them separate also makes it more obvious what is shared versus what isn't.

File: packages/dd-trace/src/collector/span.js (hunk 6 above)
Comment (rochdev) [reply packages/dd-trace/src/collector/span.js:92]: Spans have always had awareness of segments as the perhaps not greatly named `_trace` property. It _must_ know about it at the very minimum for correlation purpose.

File: packages/dd-trace/src/collector/span.js (hunk 4 above)
Comment (Qard) [reply packages/dd-trace/src/collector/span.js:36]: In both cases though you're holding the `trace.ticks` value. I'm suggesting we should try to avoid holding that in the library if at all possible. And I'm not sure why there would be drift coming from the same process on the same machine between monotonic timestamps that both came from that same process. 🤔

File: packages/dd-trace/src/collector/span.js (hunk 6 above)
Comment (Qard) [reply packages/dd-trace/src/collector/span.js:92]: We need some IDs for correlation, but we shouldn't need to hold the whole trace. Seems to me like we're not getting any of the benefits of event-based output if we're still holding all the same data for the life of the spans or traces. 🤔

File: packages/dd-trace/src/collector/span.js (hunk 4 above)
Comment (rochdev) [reply packages/dd-trace/src/collector/span.js:36]: This is also an optimization so that for applications under high load that generate traces quickly the events only need to send <=32 bits for the ticks instead of constantly 64 bits.

It also means an easier to reason about floor for the ticks which would always be 0 for the segment. However, since the number is arbitrary anyway, that doesn't actually matter too much.

Basically, it seemed to me that adding a single subtraction operation didn't add enough complexity to justify going with the alternative that saves a subtraction but ends up sending a lot more data. With the current structure of events, the timestamp is basically otherwise the largest piece of data of an event. I even considered going with microseconds since nanoseconds are rarely (read: never) useful for tracing data.

File: packages/dd-trace/src/collector/span.js (hunk 5 above)
Comment (rochdev) [reply packages/dd-trace/src/collector/span.js:18]: The parent calling a protected method on the child is how most of the tracer works since it comes from an OpenTracing background and OT does that extensively, so I don't think it should be surprising to anyone reading the code.

With that said, I can change it if you really want, but I'm not sure of the best approach then. I don't necessarily want to do it in the same file, because one of the ideas here was to not load this code at all when it's not needed. Not doing this is currently a huge problem in the library especially for serverless, and I didn't want to exacerbate the issue. What approach would you suggest with that in mind?

File: packages/dd-trace/src/collector/span.js (hunk 6 above)
Comment (rochdev) [reply packages/dd-trace/src/collector/span.js:92]: > We need some IDs for correlation, but we shouldn't need to hold the whole trace. Seems to me like we're not getting any of the benefits of event-based output if we're still holding all the same data for the life of the spans or traces.

All it's keeping is the IDs, plus a few integers used only for state tracking. It doesn't hold any actual trace data, only what is needed to know the current status of the segment, so I fail to understand your point.

File: packages/dd-trace/src/collector/span.js (hunk 4 above)
Comment (Qard) [reply packages/dd-trace/src/collector/span.js:36]: But you're trading an extra potential maximum of 32 bits of data over the wire for _more than that_ held in process memory for the duration of that span. That adds up quickly when there are potentially hundreds of thousands of concurrent spans. Process memory is a lot more scarce than bandwidth.

More concerning to me though is that we need a place to _store_ any data like this which is held in-process for the span duration. I'm _trying_ to get us _away_ from storing the big bag of data that is SpanContext, but things like this further entangle it.

File: packages/dd-trace/src/collector/span.js (hunk 5 above)
Comment (Qard) [reply packages/dd-trace/src/collector/span.js:18]: Not going to block on this one, just wanted to point out it's really hard to follow so it would be _ideal_ if we could make it more coherent somehow.

One possibility to enable only loading when you _need_ it is multiple layers of diagnostics channel. You could have the existing class publish events with subscribers that do the transformations these functions are doing and then publish the result to those collector channels. Not sure that would be any clearer, but it _would_ enable optional loading.

We could also consider some sort of dependency injection type thing where we register export strategies on the tracer and the span class could just loop over registered strategies and call each. Then the flow goes in a more singular direction rather than this confusing circular dependency.

File: packages/dd-trace/src/collector/span.js (hunk 4 above)
Comment (rochdev) [reply packages/dd-trace/src/collector/span.js:36]: It's actually the opposite. It trades the memory that would _otherwise be held in memory_ for every span because the data is buffered, to a single integer that is held only once for the whole segment. So this actually _saves_ memory in-process, and also has the benefit of reducing bandwidth. You could argue it increases CPU usage a tiny bit by adding a subtraction, but that's minimal enough to justify the trade-off.

File: packages/dd-trace/src/collector/span.js (hunk 5 above)
Comment (rochdev) [reply packages/dd-trace/src/collector/span.js:18]: These sound like great ideas to explore, but in my opinion are out of the scope of this PR. Internals are already loaded in various ways that can end up being hard to follow, and standardizing on a way to do this would be more beneficial than introducing yet another way to load things for this PR specifically.

File: packages/dd-trace/src/collector/span.js (hunk 6 above)
Comment (rochdev) [reply packages/dd-trace/src/collector/span.js:92]: That is a temporary implementation detail while both approaches have to continue working in parallel. A lot of the code relies on random things from the SpanContext today. The idea of the PR is to at least enable reporting the data as events, not to solve all the underlying problem with the rest of the code that prevent moving the state out entirely. For the events themselves however, only the IDs and active span count are needed. Removing most of the state from the process entirely will likely be several months or even quarters of work.

File: packages/dd-trace/src/collector/span.js (hunk 5 above)
Comment (Qard) [reply packages/dd-trace/src/collector/span.js:18]: I'm more concerned that piling on more complexity will make it even harder to untangle in the future whenever we get to cleaning that up. The present existence of tech debt doesn't justify making more. 😅

File: packages/dd-trace/src/collector/encoder.js [hunk 7]
```diff
@@ -0,0 +1,585 @@
...
+    return buffer
+  }
+
+  flush (done = () => {}) {
+    try {
+      const data = this.makePayload()
+
+      this._timer = clearTimeout(this._timer)
+
+      collector.send_events(data)
```
Comment (juan-fernandez) [inline packages/dd-trace/src/collector/encoder.js:208]: ❓ is `collector.send_events` sync? Do we need a `done` callback then? Can't we just return? Or is this done to preserve the `flush` signature from other exporters?

File: packages/dd-trace/src/collector/span.js (hunk 4 above)
Comment (rochdev) [reply packages/dd-trace/src/collector/span.js:36]: I still don't understand where the problem is. Correct me if I'm wrong, but these seem to be the 2 approaches we are discussing:

Option 1:
- Send timestamp and ticks for segment.
- Send raw ticks for the span.

Option 2:
- Send timestamp for segment.
- Send computed ticks for span.

For option 1, you end up having to always send 64bit values. For option 2, you often need less than 64bit since the values will start from 0 instead of the ticks when the segment started. So option 2 uses less memory, at the cost of slightly more CPU for a subtraction per span, but that is an extremely low cost, and would be offset anyway by the need to write less data to the buffer.

I don't understand how putting _less data_ in the buffer is supposed to result in _more memory_? Can you clarify your calculation?

Comment (Qard) [reply packages/dd-trace/src/collector/span.js:36]: Computed ticks requires holding additional data on the Span Context between all the captured points in the segment, while just using raw ticks everywhere does not. It also makes it more complicated to take a more stream-oriented approach with less segment buffering as the timestamp the computed ticks are in relation to changes and would therefore add additional complexity to translating them back to absolute timestamps for backend processing and frontend visualization.

There's not a whole lot of benefit to being event-based if we're going to just continue buffering all the data in-process anyway. We're just moving held memory around rather than actually getting it _out_ of the process which is the whole _point_ of wanting to go for an event model.
